src/functions.rs

```diff
@@ -2040,10 +2138,29 @@ impl ValidatedCall {
             .into());
         }
 
-        // Each argument must match the type the function signature expects
+        if let Some(cast_name) = &call.cast_name {
+            if let Some(class_name) = &call.class_name {
+                validate_cast(
+                    class_name,
+                    types.get(cast_name.as_ref()),
+                    Some(call),
+                    types,
+                    context,
+                    Some(file),
+                )?;
+            }
+        }
+
         let mut args = match &call.class_name {
-            Some(c) => vec![convert_class_name_if_this(c, parent_type)?.get_cil_name()],
-            None => Vec::new(),
+            Some(class_name) => {
+                vec![convert_class_name_if_this(class_name, parent_type)?.get_cil_name()]
+            }
+            None => match &call.cast_name {
```

Comment:
I feel like this section of code is a little inconsistent about what it assumes about the broader codebase.  I *think* that based on the current state of ast.rs it's impossible at the parser level for class_name to be None while cast_name is Some().  The block starting on 2141 seems to assume that, by not handling the class_name None case, which is reasonable enough.  But now we have special handling for is class_name is None and cast_name is Some(), which I justified the above block as being fine because that's impossible.

I think it is impossible for class_name to be None while cast_name is Some(), but I think this function should either 1. Assume that and act accordingly or 2. Verify that and return an internal error if it's false.   If 1, that should probably be documented in a comment.